### PR TITLE
Data Annotations Part 3

### PR DESCRIPTION
**Pull request by smitpatel**
part of #107 
Add support for navigation added conventions
NotMapped attribute for properties/navigations
Required attribute for navigations

### CODE REVIEW

File: src/EntityFramework.Core/Metadata/Conventions/Internal/RequiredNavigationAttributeConvention.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class RequiredNavigationAttributeConvention : NavigationAttributeConvention<RequiredAttribute>
+    {
+        public override InternalRelationshipBuilder Apply(InternalRelationshipBuilder relationshipBuilder, Navigation navigation, RequiredAttribute attribute)
+        {
+            Check.NotNull(relationshipBuilder, nameof(relationshipBuilder));
+            Check.NotNull(navigation, nameof(navigation));
+            Check.NotNull(attribute, nameof(attribute));
+
+            return relationshipBuilder.Required(true, ConfigurationSource.DataAnnotation) ?? relationshipBuilder;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Conventions/Internal/RequiredNavigationAttributeConvention.cs]: I think we should only do this if the [Required] is on a reference navigation, putting it on a collection navigation doesn't have the same semantics.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 2]
```diff
@@ -285,15 +285,17 @@ private bool CanRemove(ForeignKey foreignKey, ConfigurationSource configurationS
                 ? this
                 : ModelBuilder.Entity(foreignKey.DeclaringEntityType.Name, ConfigurationSource.Convention);
 
+            var builder = fkOwner._relationshipBuilders.Value.TryGetValue(foreignKey, ConfigurationSource.Convention) ?? fkOwner.Relationship(foreignKey, true, ConfigurationSource.Convention);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: you should be able to just call `fkOwner.Relationship(foreignKey, true, ConfigurationSource.Convention)`

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 3]
```diff
@@ -312,17 +314,23 @@ private bool CanRemove(ForeignKey foreignKey, ConfigurationSource configurationS
...
             {
-                if (_ignoredProperties.HasValue)
+                if (_ignoredProperties.HasValue && _ignoredProperties.Value.ContainsKey(navigationName))
                 {
+                    ConfigurationSource existingConfigurationSource;
+                    _ignoredProperties.Value.TryGetValue(navigationName, out existingConfigurationSource);
+                    if (!configurationSource.Overrides(existingConfigurationSource))
+                    {
+                        return builder;
+                    }
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: This is redundant as it's already checked in `CanSetNavigation`

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 4]
```diff
@@ -903,19 +911,28 @@ private InternalIndexBuilder Index(IReadOnlyList<Property> properties, Configura
                     Debug.Assert(requiredSet || !isRequired.Value);
                 }
 
+                foreignKey.IsRequired = isRequired.Value;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: This shouldn't be necessary as IsRequired is calculated from the contained properties. Instead this should be changed to an Assert and if it fails then the logic needs to be fixed.

File: src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 5]
```diff
@@ -386,7 +388,7 @@ public virtual bool Ignore([NotNull] string propertyName, ConfigurationSource co
             var navigation = Metadata.FindNavigation(propertyName);
             if (navigation != null)
             {
-                if (!Navigation(null, navigation.ForeignKey, navigation.PointsToPrincipal(), configurationSource, canOverrideSameSource: false))
+                if (Navigation(null, navigation.ForeignKey, navigation.PointsToPrincipal(), configurationSource, canOverrideSameSource: configurationSource != ConfigurationSource.Explicit) == null)
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/InternalEntityTypeBuilder.cs]: I changed my mind on this, I don't think this is the right way to fix it. Ignore behavior should be the same for any configuration source. The problem here is that we don't record the ConfigurationSource for navigations, that's what should be fixed.

**Comment by AndriySvyryd** [comment]
:shipit:, Adding a separate configuration source for navigations can be done in a different commit

**Comment by smitpatel** [comment]
merged with 03fe8f97ee1ed3618ad5c0257c89d8e3aa307488

**Pull request closed without merging**